rfcs/text/0352-graph-type-system.md

```diff
@@ -0,0 +1,2579 @@
+
+3.  define a constrained type system that is able to describe any JSON structure
+4.  refine the current way of defining links between entities, encapsulating the approach within the type system
+
+## Types
+
+Underpinning this proposal is the definition of four classes of 'types'. One of these classes is an extension of the current definition of an _Entity Type_ while the other three are new classifications that are designed to improve the usability and expressiveness of _Entity Types_.
+
+### Data Types
+
+A **Data Type** describes a space of possible valid _values_. For instance, the _string_ value `"foo"` might be an instance of a `Text` Data Type, or a _float_ value `0.6` might be an instance of a `Number` Data Type ("instance of" means that the value satisfies the constraints expressed on the Data Type, or alternatively you can view it as saying the value is within the value space of the Data Type).
+
+> 💬 The underlying value's representation will depend on the implementation environment, different languages and technologies may store these differently.
+
+A Data Type is composed of the following:
+
+- a **required** `$id`, which is a globally unique identifier, where the Data Type's definition can be accessed (in most cases this will be a URL)
+- a **required** `name`, which should generally be a non-pluralized short description of the Data Type (for example: "Number" not "Numbers")
+- **optionally** a `description` to further explain the semantic meaning of the Data Type
+- a **required** definition of its possible value
+
+#### Primitive Top-Level Data Types
+
+We pre-define a set of primitive top-level Data Types.
+These can be considered the least restrictive value spaces:
+
+- `Text`
+- `Number`
+- `Boolean`
+- `Null`
+- `Object`
+- `Empty List`
+
+#### Creating new Data-Types
+
+Due to the size of this RFC, creating new data-types has been deemed out-of-scope and a [follow-up RFC](https://github.com/blockprotocol/blockprotocol/pull/355) is in the works to spec out how this could work, please consult it for plans on introducing a suite of Data Types including ones like `Date`, `Positive Number`, etc.:
+
+### Property Types
+
+A **Property Type** is a description of a named piece of data, including its possible values. A Property Type is composed of the following:
+
+- a **required** `$id`, which is a globally unique identifier, where the property type's definition can be accessed (in most cases this will be a URL)
+- a **required** `name`, which should generally be a non-pluralized description of the property (for example: "Address" not "Addresses")
+- **optionally** a `description` to further explain the semantic meaning of the property type
+- a **required** definition of its possible values
+
+**Defining the possible values of a property**
+
+#### Data Types
+
+A Property Type can have plain data as its value, which it can define by using a Data Type.
+
+**Example 1**
```

Comment:
I think we should consider it, but generally collapsible stuff is pretty messed up in GitHub markdown I can try and find some examples later but have avoided until now for that reason.

Reply:
This particular example would look something like this:

---

A Property Type can have plain data as its value, which it can define by using a Data Type.

<details><summary>Example 1</summary>
The `Favorite Quote` Property Type could define its value as being an instance of the `Text` Data Type.

- Sample data when used in (simplified view of) an Entity

  ```json
  {
    "favoriteQuote": "He who controls the spice controls the universe"
  }
  ```
</details>

<details><summary>Example 2</summary>
The `Age` Property Type could define its value as being an instance of the `Number` Data Type.

- Sample data when used in (simplified view of) an Entity

  ```json
  {
    "age": 67
  }
  ```
</details>

---

Reply:
You can see problems in the particular example. `code` styled words aren't being rendered properly. I think we might encounter other issues too.

We're also likely going to created rendered pages from the RFCs to host ourselves so for now I think we should keep things very plain markdown to make the generation easier and more consistent. I'd prefer it if we avoided inline HTML because it also decreases readability of the raw MD

Reply:
Okay